### PR TITLE
Fix some race conditions in `BanMan::DumpBanlist()`

### PR DESCRIPTION
**Pull request by hebasto**
This PR split from bitcoin/bitcoin#24097 with some additions. This makes the following switch from `RecursiveMutex` to `Mutex` a pure refactoring.

See details in commit messages.

### CODE REVIEW

File: src/banman.cpp [hunk 1]
```diff
@@ -39,6 +39,10 @@ BanMan::~BanMan()
 
 void BanMan::DumpBanlist()
 {
+    static Mutex dump_mutex;
+    TRY_LOCK(dump_mutex, dump_lock);
+    if (!dump_lock) return;
```
Comment (MarcoFalke) [inline src/banman.cpp:44]: why is is ok to refuse to dump when two threads call this function?

Comment (hebasto) [reply src/banman.cpp:44]: It is an optimization. Calling this function in the destructor guarantees that thr final state will be dumped to the disk.

Comment (MarcoFalke) [reply src/banman.cpp:44]: Then it should be mentioned in the release notes that calling the ban rpc no longer implies the stuff is written to disk and when the node crashes, the state will be lost.

**Review by w0xlt — APPROVED** [review]
tACK e06d092

This PR : 

. adds `m_banned_mutex` to prevent the `m_is_dirty` value from being read in a different state in `BannedSetIsDirty()` and `SweepBanned()`.

. gets rid of `GetBanned()` so it doesn't call `SweepBanned()` twice and sets `banmap = m_banned`. This works because the `m_banned_mutex` is already held in this operation.

. adds the same strategy used in ` DumpMempool` to avoid concurrent writing to disk, but adds an optimization to return early if a write is in progress.

. Since there is only `m_is_dirty = true;` (not `= false`) outside of `BanMan::DumpBanlist()`, the last commit makes sense to avoid inconsistent values for `m_dirty` in a race condition.

**Review by shaavan — COMMENTED** [review]
Concept ACK

Going commit wise:

1. Concurrent calls to `Write()` have been prevented by "Trying" to lock `dump_lock` and exit if locking fails. However, as @MarcoFalke pointed out, it can lead to ban rpc calls in which `DumpBanList()` could not write anything to disk, leading to potential data loss. This behavior should be appropriately documented. 
Other than this, it could also lead to writing failure when called through other functions as well. What do you think about introducing a mutex for this function that allows calling DumpBanList only when we `AssertLockNotHeld(new_mutex)`.

2. **SweepBanned** and **BannedSetIsDirty** are brought under the same lock block to prevent race conditions for `m_is_dirty`. I agree with this approach.
3. Prevented **SweepBanned** to be called twice by directly setting the value of `banmap` (= `m_banned`) instead of calling `GetBanned()`. I think this is an improvement over the current code.
4. Modifying the code to call `SetBannedSetDirty(true);` throughout the codebase prevents the risk of inconsistent value due to race conditions. I agree with this change.

**Comment by w0xlt** [comment]
Perhaps the same strategy used in `DumpMempool` woul be a safer approach in 2c77427 .

```c++
static Mutex dump_mutex;
LOCK(dump_mutex);
```

**Comment by hebasto** [comment]
Updated e06d092f475b261afc05bbec7c4aba903d1b1a4b -> 99a6b699cd650f13d7200d344bf5e2d4b45b20ac ([pr24168.01](https://github.com/hebasto/bitcoin/commits/pr24168.01) -> [pr24168.02](https://github.com/hebasto/bitcoin/commits/pr24168.02), [diff](https://github.com/hebasto/bitcoin/compare/pr24168.01..pr24168.02)).

Lock optimization dropped as an any optimization must be justified with benchmarking, at least. That was not done in this case.
So sorry for a premature optimization.

File: src/banman.cpp (hunk 1 above)
Comment (hebasto) [reply src/banman.cpp]: Not relevant since the recent [update](https://github.com/bitcoin/bitcoin/pull/24168#issuecomment-1024569483), I guess.

**Review by w0xlt — APPROVED** [review]
reACK 99a6b69

Optimization removed, but the dump behavior now is more predictable.

**Comment by jonatack** [comment]
Concept ACK

**Review by shaavan — APPROVED** [review]
ACK 99a6b699cd650f13d7200d344bf5e2d4b45b20ac

Changes since my last review:
- Changed logic of preventing concurrent write by creating a static mutex and locking it.

@hebasto 
We can move on to the optimization logic after we have some benchmark results. Until then, we can stick with the current logic, which is more reliable and predictable.

**Review by luke-jr — APPROVED** [review]
utACK

File: src/banman.cpp [hunk 2]
```diff
@@ -39,18 +40,23 @@ BanMan::~BanMan()
...
+        LOCK(m_cs_banned);
+        SweepBanned();
+        if (!BannedSetIsDirty()) return;
+        banmap = m_banned;
         SetBannedSetDirty(false);
     }
 
+    int64_t n_start = GetTimeMillis();
+    if (!m_ban_db.Write(banmap)) {
+        SetBannedSetDirty(true);
```
Comment (MarcoFalke) [inline src/banman.cpp:57]: unrelated: Is m_dirty even used on shutdown when true? Shouldn't the below log be adjusted when it fails?

**Pull request merged**